PDB.read() function with automatic handling of gzipped files

duerrsimon opened the pull request:
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

Adds support for `pdb.gz` and `cif.gz` files for PDBParser() and MMCIFParser() as discussed in #3498 

Added a new function for PDBParser and MMCIFParser that detects if file is gzipped or not and then decodes the line.

duerrsimon committed:
add support for cif.gz and pdb.gz files + unittest

duerrsimon committed:
contrib update and pdb.gz cif.gz to news

duerrsimon committed:
change gzip mode to text and remove decoding

duerrsimon commented:
Following discussion in the issue #3498 I have changed the read mode in gzip to text instead of binary and removed the decoding. 
Technically users could also directly pass `gzip.open("filename","rt")` to `PDBParser` or `MMCIFParser` but I guess for beginners it might be more user friendly if this was handled automatically by the parser.

duerrsimon committed:
add PDB.read() function and revert changes to PDBParser(), MMCIFParser()

JoaoRodrigues commented:
Hi @duerrsimon 

I will have a look at this over the weekend for sure. Thank you for the contribution!

mdehoon commented on Bio/PDB/__init__.py:
```diff
@@ -88,3 +97,67 @@
+
+    Arguments:
+     - handle    - handle to the file, or the filename as a string
+     - format    - string describing the file format. can be [pdb, cif, mmtf]
```
I would suggest to use `fmt` instead of `format`, as `format` is also a built-in function. Over time we will probably start using `fmt` instead of `format` for existing parsers like `Bio.SeqIO`; some parsers in Biopython are already using `fmt`, for example `Bio.motifs`.

duerrsimon replied:
happy to change this. I went with the same style as `SeqIO.read`. but if other parsers already use fmt I think I would be smart to also use `fmt` here if that is the long term goal.

peterjc replied:
Python tolerates using "format" as an argument or variable name, but I agree using "fmt" is better on new code.

JoaoRodrigues replied:
Agree with `fmt`. I'd go ahead and add a kwargs that you can pass to the parser. Something like this:

```python
def read(handle, fmt, id=None, **kwargs):
    ....
    parser = Parser(**kwargs)
    ...
```

That way you can write `read(myhandle, fmt="pdb", QUIET=1)` and that `QUIET` keyword gets passed to the parsers.

mdehoon commented on Bio/PDB/__init__.py:
```diff
@@ -88,3 +97,67 @@
+    Arguments:
+     - handle    - handle to the file, or the filename as a string
+     - format    - string describing the file format. can be [pdb, cif, mmtf]
+     - id        - optional id of the structure, if None header information or then base name is used
```
Do we need this capability? If a user for some reason doesn't want to use the id provided in the header information, it can always be changed after calling `PDB.read`.

JoaoRodrigues replied:
Just a quick note: this is necessary because not all files have an ID (field or otherwise).

peterjc replied:
This was always a confusing point for me using the PDB parser - can the existing parsers not pick a sensible default or cope with a blank ID? It would be nicer not to ask the user here.

JoaoRodrigues replied:
The `id` field is just a friendly string to help you identify the structure you are working on. It has no "default" value or standard value from any file format (you could parse some sort of Id but it's inconsistent). I usually set it to the basename of the file I loaded.

mdehoon commented on Bio/PDB/__init__.py:
```diff
@@ -88,3 +97,67 @@
+    if not format:
+        raise ValueError("Format required (lower case string)")
+    if format != format.lower():
+        raise ValueError("Format string '%s' should be lower case" % format)
```
Perhaps we can be more flexible here and be case-insensitive.

duerrsimon replied:
I copied this from `SeqIO.read()`. I agree that it would be more user friendly to be case insensitive and to also accept pdb.gz and cif.gz for the CIF and PDB parser.

peterjc replied:
I would stick with "pdb" even if gzipped, rather than overloading this argument with compression information.

I'm open to having the format names treated case insensitively here and potentially in SeqIO.

JoaoRodrigues replied:
Agree with case insensitivity. In addition, I'd move format checking to its own function. Basically a `_check_format` helper that takes a string and returns a parser class (not an instance) or throws an exception.

mdehoon replied:
Agree with @peterjc  to use "pdb" also for gzipped files.

mdehoon commented on Bio/PDB/__init__.py:
```diff
@@ -88,3 +97,67 @@
 except ImportError:
     pass
+
+_FormatToParser = {"pdb": PDBParser(), "cif": MMCIFParser()}
```
This uses the same parser instance when multiple files are parsed. Are `PDBParser` and `MMCIFParser` reusable?

duerrsimon replied:
As far as I can tell they are. Here in this example https://biopython.org/wiki/PDBParser the PDBParser is reused for benchmarking.

JoaoRodrigues replied:
I'd avoid this. We might want to pass arguments to the parsers via kwargs.

mdehoon commented on Bio/PDB/__init__.py:
```diff
@@ -88,3 +97,67 @@
+        # if we are provided a path to a file, check if gzipped,
+        # else open handle directly
+        if isinstance(handle, str):
+            encoding = guess_type(handle)[1]  # uses file extension
```
Here, you could uses `guess_type` in a `try: else:` block. This allows you to accept `str`, `bytes`, and `os.PathLike` objects as valid file names.

JoaoRodrigues replied:
I thought `guess_type` was actually checking the file, not just the extension. If that's the case, we might as well ignore `mimetypes.guess_type` and use `Path.suffix` to match the dictionary mapping extensions to parsers (`_FormatToParser`).

duerrsimon replied:
I know that e.g biological assemblies come in pdb1.gz format so then we would need to match just the gz part because a file might be of format pdb but have a different extension. 
`guess_type` really only looks at the filename but I though it'd be nice and readable this way. 
Alternatively one could look at the first two bytes of the file and determine if it is  gzipped but afaik there are also very few edge cases where this does not work 100% reliably. I think `gunzip` also only looks at the filename whether it is .gz extension so I think this is reasonably safe this way. 

Beginning Python3.8 there is also a gzip.BadGzipFile Exception but I guess for backwards compatibility this can only be implemented in the somewhat clunky wat: 
```
import gzip
with gzip.open(filename, 'r') as fh:
    try:
        fh.read(1)
    except OSError:
        print('input_file is not a valid gzip file by OSError')
    except gzip.BadGzipFile: # only Python 3.8+
        print('input_file is not a valid gzip file by BadGzipFile')
```

mdehoon commented on Bio/PDB/__init__.py:
```diff
@@ -88,3 +97,67 @@
+            encoding = guess_type(handle)[1]  # uses file extension
+            _open = partial(gzip.open, mode="rt") if encoding == "gzip" else as_handle
+        else:
+            _open = as_handle
```
Do we still need `as_handle` here? If `open` is sufficient, then I would prefer that is it makes it easier to follow the logic in this code section.

duerrsimon replied:
`guess_type` will fail if a handle is passed. I will change it together with the the `try: else:` block to make the logical flow better.

mdehoon replied:
I like the approach with `guess_type`. You could consider what to do if `gzip.open` fails because the file turned out not to be gzipped after all. But maybe this possibility is too remote and we can just ignore it.

duerrsimon replied:
see above. I will add a check for bad gzip files once we have decided on a way to to check if it is a gzip file ( via magic, guess_type or else).

mdehoon commented on Bio/PDB/__init__.py:
```diff
@@ -88,3 +97,67 @@
+            if structure.header["idcode"] != "" and id is None:
+                structure.id = structure.header["idcode"]
+            elif isinstance(handle, str) and id is None:
+                structure.id = Path(handle).stem
```
Perhaps we can just leave `structure.id` as `None` in this case? Getting information from the file name seems a bit unusual.

duerrsimon replied:
If a PDB file has no header it is most likely a PDB file generated by some kind of computational workflow so in that case I'd guess that the filename has some meaningful information.  So while unusual it might still be useful if loading multiple structures at the same time in order to identify the structures instead of having multiple structures with `None`.

mdehoon replied:
Wouldn't this lead to the situation that two files with identical contents but different filenames get different structures? For example, `1LCD.pdb` in the Biopython tests does not have an ID in its header. Then this could happen:
```
mdehoon ~/biopython-master/Tests/PDB 
$ cp 1LCD.pdb test.pdb

mdehoon ~/biopython-master/Tests/PDB 
$ python              
Python 3.9.1 (v3.9.1:1e5d33e9b9, Dec  7 2020, 12:44:01) 
[Clang 12.0.0 (clang-1200.0.32.27)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
```
```python
>>> from Bio import PDB
>>> s = PDB.read("1LCD.pdb", "pdb")
[... some unrelated warnings ...]
>>> t = PDB.read("test.pdb", "pdb")
[... some unrelated warnings ...]
>>> s == t
False
>>> s.id
'1LCD'
>>> t.id
'test'
>>> s.id = None
>>> t.id = None
>>> s == t
True
```

duerrsimon replied:
Yes this can happen. The id is however only descriptive as far as I am aware so in real world applications I wouldn't say that it is undesired to indicate that the original file used to construct the structure was named differently.

peterjc replied:
I'd echo @mdehoon's concern about treating the filename as meaningful. None seems a better choice.

mdehoon replied:
Actually I am wondering now what the purpose of the `id` attribute is.

If I parse a PDB file with a complete header line, like 1A8O.pdb in our tests, and I specify `"test"` as the `id`, then this is stored as `structure.id`:
```python
>>> from Bio.PDB import PDBParser
>>> parser = PDBParser()
>>> structure = parser.get_structure("test", "1A8O.pdb")
>>> structure.id
'test'
```
But the ID code from the header is also stored by the parser:
```python
>>> structure.header['idcode']
'1A8O'
```
Then what is the purpose of `id`? The `get_structure` docstring says
```
get_structure(id, file) method of Bio.PDB.PDBParser.PDBParser instance
    Return the structure.
    
    Arguments:
     - id - string, the id that will be used for the structure
     - file - name of the PDB file OR an open filehandle
```
but I could not understand the purpose of `id` from this. @JoaoRodrigues  do you know what the `id` attribute is good for?

peterjc replied:
I never understood what the user-supplied ID was for when using the PDB module.

JoaoRodrigues replied:
As I commented above, I actually agree with @duerrsimon and would go with the filename as ID. The ID is only used when printing a `Structure` (to the command line). Other than that, it has no value. IDs of other Entities are meaningful though. That said, we could keep it as None but, I think the file name is a very good default value as well. Always helps to have a point to where the structure came from.

mdehoon replied:
If`Bio.PDB` does not need the `id` attribute, I think we should not set it at all (i.e. don't even set it to `None`). We don't store the filename in other Biopython parsers. As noted above, `__eq__` on identical structures stored in different files will return `False` ; this is very confusing. If it is the same file known under a different name, e.g. a soft-linked file, comparing the structures will also return `False`. Even if it is the same file under the same name, you may get a different `id` (for example, `structure = PDB.read('myfile.pdb')` and `handle = open('myfile.pdb'); structure = PDB.read(handle)` will give structures with different IDs. And if the `id` is the same, you still won't know if the structures came from the same file: They may come from different files that happen to have the same file name but are stored in different directories, or one stored locally and the other one remotely.

I can see that in specific cases it may be useful to store the filename as an attribute to the structure, depending on the circumstances and purpose in each case, but in general, I think this will open up endless opportunities for bugs. Users who do want to store the filename as an `id` attribute can always do so after `PDB.read` returns.

Note also that the `id` attribute on components of the structure (such as the individual chains) are read from the contents of the PDB file and are well-defined. There is some inconsistency in using the same attribute name for a well-defined property and for an essentially arbitrary property.

duerrsimon replied:
Ok I get your concerns.  
So lets keep it at the following then? 
- if optional id is set use that as id
- if file contains id in header = use that as id  (will be consistent for identical files with different names)
- else set to None

Maybe it'd be worth to change __repr__ of structure so that id is only printed if set? 
printing  `<Structure>` instead of `<Structure id=None>` 
None of the unit tests fails if I implement this.

peterjc replied:
Why complicate things, does the new ``Bio.PDB.read(...)`` function need an ``id=...`` argument at all? Surely just two mandatory arguments are enough for now: filename/handle and format.

JoaoRodrigues replied:
Comparing structures using `__eq__` is a bad idea. It's not even a well-defined question. For example, to begin with, do you want to compare coordinates or chemical composition?

I'd leave it as None to be honest (which is the equivalent to being not set).

mdehoon replied:
I agree with @JoaoRodrigues  that it is unclear what `__eq__` should be doing. But currently `__eq__` is  already defined for `Structure` objects in the `Entity` class from which `Structure` inherits. This `__eq__` implementation just checks whether the `id` attribute is the same. The same `__eq__` is used for other entities like chains and atoms. This definition of `__eq__` gives some surprising results, for example chains coming from different structures may be evaluated as equal:
```
>>> from Bio.PDB import PDBParser
>>> parser = PDBParser()
>>> structure1 = parser.get_structure('1A8O', "1A8O.pdb")
>>> structure2 = parser.get_structure('1LCD', "1LCD.pdb")
>>> chains1 = list(structure1.get_chains())
>>> chains2 = list(structure2.get_chains())
>>> chains1[0].id
'A'
>>> chains2[0].id
'B'
>>> chains2[1].id
'C'
>>> chains2[2].id
'A'
>>> chains1[0] == chains2[2]
True
```
Then I think it's fair to say that `__eq__` in `Bio.PDB` is broken.

I don't think `id` should be an argument to the `read` function, not even with default value `None`, because doing so will make it part of the official API, and it takes time to change that (having to declare it obsolete, and then two releases with a deprecation warning on it before we can change it).

To move forward, how about the following:
- define `read` as ` def read(handle, format):` without an `id` argument;
- if the file contains `id` in the header, use that as`id`;
- else set `id` to None;
- deprecate `__eq__` as well as accessing (but not setting) the `id` attribute.

peterjc replied:
I agree with @mdehoon except on the very final point of deprecating accessing the ``id`` - it may be useful for those files which contain it.

mdehoon replied:
@peterjc  For those files, the ID is still accessible as `structure.header['idcode']`.
But I am OK to keep the `id` attribute as is for now, and discuss whether the `id` attribute should be deprecated in a separate PR.

peterjc replied:
Appreciate the clarification. I'd remove any ID handling from the new ``read`` function (and leave the existing parsers to set it if they do currently).

JoaoRodrigues replied:
Just to clarify: `__eq__` is not broken. It's comparing labels of different components so it makes sense that if two different proteins chains have the same `id` they are equal. We never intended (afaik) to use `__eq__` for true equivalency between chemical structures, either coordinate or composition based. There is a lot of modeling in protein work so structures often don't have an "official" id, just the one you (user) assign them. Exceptions are, as you found, structures downloaded from the PDB, which do have an ID code. But even there there is redundancy (multiple PDB IDs point to the same protein and conformation). Also, I can tell you that in more than 10 years using Biopython, I _never_ used `structure.header["idcode"]` :)

The `Structure.id` is a by-product of `Entity` and inheritance. While you could argue that it's hard to define what it should be, compared to a `Chain.id` for example, and that it could be removed, it's still useful in practice for many cases. So, having a way to define when you load the structure would be handy.

Finally, I don't think a new function should introduce a new API that differs from the underlying code we already have. While we might want to move on to greener pastures, this is definitely a minor point that will only confuse our users. In the future, if people use `read` by default, we can rework the parsers behind the scenes and simply have deprecation warnings for these arguments as we remove them.

For these reasons, I'd leave the `id` in as an optional argument, with a default value of`None`.

mdehoon replied:
Since we haven't reached an agreement here, and we should not keep @duerrsimon waiting forever, let's defer the discussion of `__eq__` and the purpose and usage of the `id` attribute to a separate PR.

@peterjc 's suggestion:

> I'd remove any ID handling from the new read function (and leave the existing parsers to set it if they do currently).

then seems reasonable.

peterjc replied:
OK with you @JoaoRodrigues?

It sounds like you have some real world use cases for setting the top level ID (something I have never needed to do while working with structures), which may argue for _later_ including this as an optional argument to the new ``Bio.PDB.read(...)`` function?

JoaoRodrigues replied:
I'd be fine with `Bio.PDB.read()` setting `id` to `None` internally but not exposing it as an optional argument for now!

mdehoon commented on NEWS.rst:
```diff
@@ -106,6 +106,9 @@ unambiguously guessed from the atom name, in accordance with PDB structures.
 centers of gravity or geometry.
 
+Bio.PDB.PDBParser and Bio.PDB.MMCIFParser now also read `pdb.gz` and `cif.gz`
+files.
```
I believe that that was already the case?

duerrsimon replied:
Yeah I forgot to take this out. I will change the news and add the remaining docs when we have all agreed that the pull request is ready.

duerrsimon committed:
added PDB.read mmtf tests to test_mmtf and fix doctest

duerrsimon commented:
I just fixed the doctest that was still failing and I added the tests for the PDB.read() with the mmtf parser to `test_mmtf` so that they run only when mmtf is available.

duerrsimon commented:
I also changed the title of the Pull request to reflect the changes to the implementation discussed in #3498

Lucioric2000 commented:
Great to make id optional. Then, the file name should be the first argument of the PDB parser function.

JoaoRodrigues commented on Bio/PDB/__init__.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright (C) 2002, Thomas Hamelryck (thamelry@binf.ku.dk)
+# Copyright (C) 2002,2020 Thomas Hamelryck (thamelry@binf.ku.dk), Simon Duerr
```
I would leave the copyright notice as is. We've had many contributions and it'd be a very long line if we included everyone's name. That's why we have the `CONTRIBUTORS.md` file and also why I wrote the list of "major" contributors below (in alphabetical order).

peterjc replied:
And when we do have multiple authors listed, we typically have one name per line. e.g.

```
# Copyright (C) 2002, Thomas Hamelryck (thamelry@binf.ku.dk)
# Copyright (C) 20201, Simon Duerr
```

This scales better, and makes things like merge conflicts less likely and easier to resolve.

JoaoRodrigues commented on Bio/PDB/__init__.py:
```diff
@@ -19,15 +19,29 @@
 - Joao Rodrigues
 - Kristian Rother
 - Eric Talevich
+- Simon Duerr
```
This list is in alphabetical order of surnames.

JoaoRodrigues commented on Bio/PDB/__init__.py:
```diff
@@ -19,15 +19,29 @@
+    pass
+
+
+from Bio.File import as_handle
```
Although this is not a "rule", I usually order imports as: standard lib, third-party, intra-package, intra-module. So, I'd move this up before the parsers.

JoaoRodrigues commented on Bio/PDB/__init__.py:
```diff
@@ -88,3 +102,67 @@
+    # need to use different parser for mmtf versus pdb,cif
+    if format == "mmtf":
+        # TODO: Raise warning if ID is set. MMTFParser does not support passing an id
+        # MMTF needs a seperate parser as get_structure does only take 1 positional argument and does not support file handles
```
I'd ignore `id` indeed. Alternatively, we can add a dummy id field to that `get_structure` so that we have a unified interface (easier than adding logic on this end).

JoaoRodrigues commented on Tests/test_PDB.py:
```diff
@@ -0,0 +1,88 @@
+# Copyright 2021 by Simon Duerr.  All rights reserved.
```
I would move these tests to the respective parsers test files.

duerrsimon replied:
But where would the tests for the fmt string and so on go? They are not related to the parsers directly right?  
To me it is logical to have the tests that test the read function itself in `test_PDB` and to have the tests that check the actual parsers e.g in `test_PDB_PDBParser` etc.

JoaoRodrigues replied:
They are somewhat directly related to the parsers, but I see your point. Let's keep them here.

JoaoRodrigues commented on Tests/test_mmtf.py:
```diff
@@ -12,13 +12,49 @@
 import warnings
 import os
 import tempfile
+from Bio import PDB
```
You can follow the convention below and simply do `from Bio.PDB import read`

duerrsimon replied:
But isn't that inconsistent with other subpackages like `SeqIO` where always SeqIO.read(..) is used?

peterjc replied:
Well, "read" on its own is a very ambiguous function name so ``SeqIO.read`` or ``AlignIO.read`` etc make much more readable code. I would recommend using ``BDB.read`` in our documentation for the same reason.

However, if the rest of ``test_mmtf.py`` uses the other style, that's fine.

JoaoRodrigues replied:
👍

JoaoRodrigues commented on Tests/test_mmtf.py:
```diff
@@ -12,13 +12,49 @@
+        # MMTF reader does not accept ids, so a warning should show that it id is ignored
+        _ = PDB.read("PDB/1A8O.mmtf", format="mmtf", id="test")
+
+    def test_PDB_read(self):
```
I'd change the name of the test (and the docstring) to be more explicit:

```python
def test_PDB_read_MMTF(self):
   """Test that read(fmt="mmtf") returns a valid structure."""
```

JoaoRodrigues reviewed (changes requested):
Overall, very nice work @duerrsimon. I think with a few tweaks this will be a very nice contribution!

JoaoRodrigues commented on Tests/test_PDB.py:
```diff
@@ -0,0 +1,88 @@
+    """Base class for Bio.PDB unit tests."""
+
+    def setUp(self):
+        self.PDBParser = PDBParser()
```
I'd rename these to `self.pdb_parser` and `self.mmcif_parser` instead of the uppercase variable names.

peterjc replied:
Good point, in line with PEP8 conventions about classes vs variables and attributes.

JoaoRodrigues commented on Tests/test_PDB.py:
```diff
@@ -0,0 +1,88 @@
+    def setUp(self):
+        self.PDBParser = PDBParser()
+        self.MMCIFParser = MMCIFParser()
+        pass
```
Not necessary here.

JoaoRodrigues commented on Tests/test_PDB.py:
```diff
@@ -0,0 +1,88 @@
+        structure = PDB.read("PDB/1A8O.pdb", format="pdb")
+        structure_gzip = PDB.read("PDB/1A8O.pdb", format="pdb")
+
+        with gzip.open("PDB/1A8O.pdb.gz", mode="rt") as handle:
```
I would split this up in two tests: read_pdb and read_pdbgz.

JoaoRodrigues commented on Tests/test_PDB.py:
```diff
@@ -0,0 +1,88 @@
+        self.assertEqual(structure, structure_handle)
+        self.assertEqual(structure, structure_parser)
+
+    def test_PDB_read_format(self):
```
I would include a test where format is missing.

JoaoRodrigues commented on Tests/test_PDB.py:
```diff
@@ -0,0 +1,88 @@
+        with self.assertRaises(TypeError):
+            _ = structure = PDB.read("PDB/1A8O.pdb", format=0)
+
+        with self.assertRaises(ValueError):
```
Remove this to be consistent with the changes proposed elsewhere (to make `format` case-insensitive).

duerrsimon committed:
implement changes to PDB.read() regarding id and reformat PDB.read unit tests

duerrsimon commented:
So I think I have implemented everything that was discussed above:

- format uppercase or lowercase
- rename format to fmt
- add `_check_format` helper function + allow kwargs for the parsers
- allow str, PathLike, Bytes or handle as input  (had to abandon guess_types because it supports PathLike only in 3.8+)
- add helpful message in case structure cannot be parsed because of wrong specified fmt (invalid string literal ... is not so helpful imho)

There should corresponding unittests for all the changes above. All the test for the read function with very basic structure comparisons reside in `test_PDB`, all further tests for correct structure parsing are with the parsers. 

Sorry for taking a bit of time on this.

mdehoon commented on Bio/PDB/__init__.py:
```diff
@@ -88,3 +98,84 @@
+    if fmt in _FormatToParser:
+        return _FormatToParser[fmt](**parseropt)
+    else:
+        raise ValueError("Unknown format '%s'" % fmt.lower())
```
I think `lower` is not needed here.

duerrsimon committed:
fmt.lower() fixes

duerrsimon commented:
Any feedback on this @JoaoRodrigues ? I am happy to also implement a PDB.write() and PDB.fetch() function (perhaps in a seperate pull request) if you want to unify the the interface a bit more  :)

JoaoRodrigues commented:
Hi @duerrsimon 

Definitely will give you feedback soon! Sorry for the delay - a lot happening at the same time. If you want, go ahead and work on the write/fetch methods in separate PRs, I'd be happy to work on them with you/review them!

JoaoRodrigues commented:
Hi @duerrsimon ,

Sorry for the delay but these have been tough months to have time to spare. I had a look at the PR and it is in great shape. I added some logic to parse the format from the file name (or raise an error), to make it simpler to use `PDB.read("1xyz.pdb")`. You can still force the format if you want.

I reviewed your tests too, some were failing and I think there was some unnecessary parsing going on. Since the changes were a little too much, I preferred to push to a branch on my own fork that you can review and cherry-pick from. Let me know what you think and again, thanks for the patience and the code! This will be a very useful piece of code, thus the long review and long list of changes :)

https://github.com/JoaoRodrigues/biopython/commit/94d59686e1454e48f0ffe5cfba5e0667d7319d6f
